src/hotspot/share/opto/output.cpp

```diff
@@ -974,6 +974,27 @@ bool PhaseOutput::contains_as_owner(GrowableArray<MonitorValue*> *monarray, Obje
   return false;
 }
 
+// Determine if there is a scalar replaced object description represented by 'ov'.
+bool PhaseOutput::contains_as_scalarized_obj(JVMState* jvms, MachSafePointNode* sfn,
+                                             GrowableArray<ScopeValue*>* objs,
+                                             ObjectValue* ov) const {
+  for (int i = 0; i < jvms->scl_size(); i++) {
+    Node* n = sfn->scalarized_obj(jvms, i);
+    // Other kinds of nodes that we may encounter here, for instance constants
+    // representing values of fields of objects scalarized, aren't relevant for
+    // us, since they don't map to ObjectValue.
+    if (!n->is_SafePointScalarObject()) {
+      continue;
+    }
+
+    ObjectValue* other = (ObjectValue*) sv_for_node_id(objs, n->_idx);
```

Comment:
Is the cast here necessary? I see them generally in the file... but not sure why.
```
ObjectValue*
PhaseOutput::sv_for_node_id(GrowableArray<ScopeValue*> *objs, int id) {
```

Reply:
Because `objs` array may contains 3 types of objects: `ScopeValue`, `ObjectValue` and `ObjectMergeValue`.
I would leave this code as it is for this PR but suggest to file followup RFE to clean this up.
Instead of such casts we should use `as_ObjectValue()` and `as_ObjectMergeValue()` which have asserts to check type.

Reply:
Thank you @eme64 , @vnkozlov for reviewing. I created this RFE to remove the unnecessary casts: https://bugs.openjdk.org/browse/JDK-8336495